pkg/querier/block.go

```diff
@@ -80,91 +124,227 @@ func (b *BlockQuerier) Get(ctx context.Context, userID string, from, through mod
+	sort.Slice(chunks, func(i, j int) bool {
+		return chunks[i].MinTime < chunks[j].MinTime
+	})
+
+	return &blockQuerierSeries{labels: labels, chunks: chunks}
+}
+
+type blockQuerierSeries struct {
+	labels []storepb.Label
+	chunks []storepb.AggrChunk
+}
+
+func (bqs *blockQuerierSeries) Labels() labels.Labels {
+	res := make(labels.Labels, 0, len(bqs.labels))
+	for _, l := range bqs.labels {
 		// We have to remove the external label set by the shipper
-		if label.Name == tsdb.TenantIDExternalLabel {
+		if l.Name == tsdb.TenantIDExternalLabel {
 			continue
 		}
-
-		lbls = append(lbls, labels.Label{
-			Name:  label.Name,
-			Value: label.Value,
-		})
+		res = append(res, labels.Label{Name: l.Name, Value: l.Value})
 	}
+	return res
+}
 
-	chunks := make([]chunk.Chunk, 0, len(series.Chunks))
+func (bqs *blockQuerierSeries) Iterator() storage.SeriesIterator {
+	if len(bqs.chunks) == 0 {
+		// should not happen in practice, but we have a unit test for it
+		return errIterator{err: errors.New("no chunks")}
+	}
 
-	for _, c := range series.Chunks {
-		ch := encoding.New()
+	its := make([]chunkenc.Iterator, 0, len(bqs.chunks))
 
-		enc, err := chunkenc.FromData(chunkenc.EncXOR, c.Raw.Data)
+	for _, c := range bqs.chunks {
+		ch, err := chunkenc.FromData(chunkenc.EncXOR, c.Raw.Data)
 		if err != nil {
-			return nil, errors.Wrap(err, fmt.Sprintf("failed to initialize chunk from XOR encoded raw data (series: %v min time: %d max time: %d)", lbls, c.MinTime, c.MaxTime))
+			return errIterator{err: errors.Wrapf(err, "failed to initialize chunk from XOR encoded raw data (series: %v min time: %d max time: %d)", bqs.Labels(), c.MinTime, c.MaxTime)}
 		}
 
-		it := enc.Iterator(nil)
-		for it.Next() {
-			ts, v := it.At()
-			overflow, err := ch.Add(model.SamplePair{
-				Timestamp: model.Time(ts),
-				Value:     model.SampleValue(v),
-			})
-			if err != nil {
-				return nil, errors.Wrap(err, fmt.Sprintf("failed adding sample to chunk (series: %v timestamp: %d value: %f)", lbls, ts, v))
-			}
-
-			if overflow != nil {
-				chunks = append(chunks, chunk.NewChunk(userID, client.Fingerprint(lbls), lbls, ch, model.Time(c.MinTime), model.Time(c.MaxTime)))
-				ch = overflow
-			}
-		}
+		it := ch.Iterator(nil)
+		its = append(its, it)
+	}
 
-		// Ensure the iteration has not been interrupted because of an error
-		if it.Err() != nil {
-			return nil, errors.Wrap(it.Err(), fmt.Sprintf("failed reading sample from encoded chunk (series: %v min time: %d max time: %d)", lbls, c.MinTime, c.MaxTime))
-		}
+	return &blockQuerierSeriesIterator{series: bqs, iterators: its}
+}
 
-		if ch.Len() > 0 {
-			chunks = append(chunks, chunk.NewChunk(userID, client.Fingerprint(lbls), lbls, ch, model.Time(c.MinTime), model.Time(c.MaxTime)))
+// blockQuerierSeriesIterator implements a series iterator on top
+// of a list of time-sorted, non-overlapping chunks.
+type blockQuerierSeriesIterator struct {
```

Comment:
Honestly, it doesn't seem worth the troubles for 50 lines of code to me.

But since I just found a bug in that iterator code (it calls `At()` before calling `Next()`, which it shouldn't), I may try to send patch with the bugfix and making it public. (But it's still low prio)